### PR TITLE
Add solution for assignment 3

### PR DESCRIPTION
**Pull request by lizaku**
One file is with solution itself, the other one is with tests

### CODE REVIEW

File: lizaku/assignment3/assignment3.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
+class Dictionary:
+    '''
+    I implement the dictionary as a trie. As an input I give an array of words
```
Comment (ariava) [inline lizaku/assignment3/assignment3.py:3]: Good job explaining your implementation choices!

File: lizaku/assignment3/assignment3.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
+class Dictionary:
+    '''
+    I implement the dictionary as a trie. As an input I give an array of words 
+    that are then transformed into a trie. isWord and isPrefix are methods of the Dictionary.
+    '''
+    def __init__(self, data):
+        self.data = self._make(data)
+    
+    def _make(self, words):
```
Comment (ariava) [inline lizaku/assignment3/assignment3.py]: Why did you separate this in a different method? Couldn't it be directly in the __init__()?

File: lizaku/assignment3/assignment3.py [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+class Grid:
+    '''
+    Grid is initially a 2D array, but I turn all values into Node objects
+    and augment them with their neighbors
+    '''
+    def __init__(self, data):
+        self.data = self._make(data)
+    
+    def _make(self, data):
```
Comment (ariava) [inline lizaku/assignment3/assignment3.py]: Why did you separate this in a different method? Couldn't it be directly in the init()?

File: lizaku/assignment3/assignment3.py [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+        root = dict()
+        for word in words:
+            current_dict = root
+            for letter in word:
+                current_dict = current_dict.setdefault(letter, {})
+            current_dict[_end] = _end
+        return root
+    
+    def isWord(self, word):
+        _end = '_end_'
```
Comment (ariava) [inline lizaku/assignment3/assignment3.py]: Just a note (no change required): here you use the string '_end_' as a marker. What you could do if you wanted to allow '_end_' as a value could be to have a specific Node object for each level of the trie and have the Node keep an attribute like is_word or is_final that is set to True only if you have reached the end of a full word.

File: lizaku/assignment3/assignment3.py [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        '''
+        I look for all possible strings in the grid via depth-first search. 
+        This is done for every element in the grid individually.
+        '''
+        if vocab.isPrefix(start.val):
+            # maybe the value of a node is a valid word
+            yield start.val
+        stack = [(start, [start])]
+        while stack:
+            (node, path) = stack.pop()
```
Comment (ariava) [inline lizaku/assignment3/assignment3.py]: I  think you can drop the parentheses around "node, path".

**Review by ariava — CHANGES_REQUESTED** [review]
Good job! Please have a look at my comments.

File: lizaku/assignment3/assignment3.py (hunk 2 above)
Comment (lizaku) [reply lizaku/assignment3/assignment3.py]: I just have a stereotypic thinking that some major transformations shouldn't be in the __init__ function... Also, if I do several things with the data that I pass, it should be different functions, right?

File: lizaku/assignment3/assignment3.py (hunk 3 above)
Comment (lizaku) [reply lizaku/assignment3/assignment3.py]: But I moved this method to init(), if you think that is more accurate.

File: lizaku/assignment3/assignment3.py (hunk 4 above)
Comment (lizaku) [reply lizaku/assignment3/assignment3.py]: Yes, but I don't make a proper class for Trie here, though it would be a better decision.

**Comment by lizaku** [comment]
Hello @ariava! Thank you for your comments, I addressed them in the fixes. Can you check it?

**Review by ariava — APPROVED** [review]
Awesome!

**Pull request merged**